### PR TITLE
Add new config_file_service_registration token

### PR DESCRIPTION
**Pull request by pglass**
### Description

This adds a new agent token: `config_file_service_registration`. This token is used to register services and checks that are defined in local config files (including when defined in flags, such as with `-hcl`).

This adds:

* The config field `acl.tokens.config_file_service_registration`
* The `PUT /agent/token/config_file_service_registration` HTTP API request
* The `consul acl set-agent-token config_file_service_registration <token>` command

The precedence of tokens when registering a service from a service definition or a check from a check definition is:

1. **Inline service token**:  The token from the `token` field in the service/check definition is used, if set
2. **Config File Service Registration token**: otherwise, the config file registration token is used, if set
3. **Default token**: otherwise, the default token is used, if set
4. **Anonymous token**: otherwise, the anonymous token is used

### Testing & Reproduction steps

- Updated unit tests
- Also, manually tested:

  - [x] Setting `acl.tokens.config_file_service_registration` and seeing successful registration of a service definition
  - [x] Setting `acl.tokens.config_file_service_registration` and seeing successful registration of a check definition
  - [x] Defining an inline token in the service definition and validating that token is used instead of the `acl.tokens.config_file_service_registration`
   - [x] Defining an inline token in the check definition and validating that token is used instead of the `acl.tokens.config_file_service_registration`
  - [x] Unsetting `acl.tokens.config_file_service_registration` and setting `acl.tokens.default` to check that checks and services fall back to the default token
  - [x] Registering services via the HTTP ensure that the `config_file_service_registration` token is only used for registering services sourced from config files
  - [x] Running `consul acl set-agent-token config_file_service_registration` and checking that `<data-dir>/acl-tokens.json` is updated when token persistence is enabled, and that the updated token is used for subsequent service registrations

### Links

https://github.com/hashicorp/consul/issues/4478

### PR Checklist

* [x] updated test coverage
* [x] external facing docs updated
* [ ] not a security concern

### CODE REVIEW

File: agent/local/state.go [hunk 1]
```diff
@@ -275,20 +284,21 @@ func (l *State) addServiceLocked(service *structs.NodeService, token string) err
...
+		IsLocallyDefined: isLocal,
 	})
 	return nil
 }
 
 // AddServiceWithChecks adds a service entry and its checks to the local state atomically
 // This entry is persistent and the agent will make a best effort to
 // ensure it is registered
-func (l *State) AddServiceWithChecks(service *structs.NodeService, checks []*structs.HealthCheck, token string) error {
+func (l *State) AddServiceWithChecks(service *structs.NodeService, checks []*structs.HealthCheck, token string, isLocal bool) error {
```
Comment (pglass) [inline agent/local/state.go:297]: A todo is to figure out if registering checks should also use the config_file_registration token. I imagine probably so, but I haven't thought it through yet.

File: agent/local/state.go [hunk 2]
```diff
@@ -1362,9 +1372,23 @@ func (l *State) pruneCheck(id structs.CheckID) {
 	delete(l.checks, id)
 }
 
+// RegistrationTokenFallback returns a fallback function to be used when
+// determining the token to use for service sync.
+//
+// The fallback function will return the config file registration token if the
+// given service was sourced from a service definition in a config file.
+func (l *State) RegistrationTokenFallback(key structs.ServiceID) func() string {
```
Comment (kisunji) [inline agent/local/state.go:1380]: nit: I don't think this is used outside `package state` so it could be private.

Alternatively, what do you think about merging this logic into the body of `aclTokenForServiceSync` since it already does a lookup of `l.services[key]`?

Passing around a lock-guarded map in a closure makes me a little cautious; in this PR the codepaths are synchronized but this could be accidentally misused in the future.

File: agent/token/store.go [hunk 3]
```diff
@@ -216,6 +224,27 @@ func (t *Store) UpdateReplicationToken(token string, source TokenSource) bool {
 	return changed
 }
 
+// UpdateConfigFileRegistrationToken replaces the current config file registration token
+// in the store. Returns true if it was changed.
+func (t *Store) UpdateConfigFileRegistrationToken(token string, source TokenSource) bool {
+	return t.updateToken(token, source,
+		&t.configFileRegistrationToken, &t.configFileRegistrationTokenSource,
```
Comment (kisunji) [inline agent/token/store.go:231]: Since this is a method off `(t *Store)` anyways, can we omit `dstToken *string, dstSource *TokenSource` from the signature and use the `t.configFileRegistrationToken` and `t.configFileRegistrationTokenSource` fields directly in `updateToken`?

**Review by kisunji — APPROVED** [review]
Left some minor comments but LGTM. What do you think about renaming `config_file_registration` to something shorter like `registration` or `static_registration`? It feels a little verbose but maybe the detailed name makes it more clear.

**Comment by jkirschner-hashicorp** [comment]
> What do you think about renaming config_file_registration to something shorter like registration or static_registration? It feels a little verbose but on the other hand maybe the detailed name makes it more clear.

In my experience, Consul agent token names are commonly misunderstood, such as `agent` and `default`. I personally prefer that we have an unambiguous name that is 3 words (`config_file_registration`). I feel like `registration` is too ambiguous, as HTTP API calls are also methods to perform "registration", but wouldn't use this token. `static_registration` resolves that ambiguity, though because config files are the only way to perform static registration (AFAIK), using `config_file_` rather than `static_` seems best to me, as there's no interpretation required on the part of the user (to figure out what "static" refers to in this context).

File: agent/local/state.go (hunk 2 above)
Comment (pglass) [reply agent/local/state.go:1380]: > I don't think this is used outside package state so it could be private.

I agree. I had to export the method because the test package is different (`local` vs `local_test`). Which seems unusual to me. But, without exporting it I can't call it in unit tests.

> Alternatively, what do you think about merging this logic into the body of aclTokenForServiceSync since it already does a lookup of l.services[key]?

I opted against this because `aclTokenForServiceSync` is also used in [deleteService](https://github.com/hashicorp/consul/blob/main/agent/local/state.go#L1269-L1274).

So, do we want deleteService to also incorporate the config file registration token in its list of fallback tokens? Generally, it seems better to me if it does not.

The main concern to me is if the config_file_registration token has been deleted, then it would fail to deregister the service and we'd see errors in logs. Also, it should fallback to using the agent token anyway. (It is able to use the agent token for service deregistrations because the Catalog.Deregister RPC accepts a token with the relevant node:write permissions).

Because the agent token must have node:write permissions (or else it could not have registered it's node into the catalog) and because the agent token is probably less likely to have been deleted (because agent lifecycle is longer than service instance lifecycle), it seems like the agent could skip straight to using the agent token without considering the config_file_registration token for the deregistration, rather than incorporating the agent token in the list of "fallback" tokens. And that would be faster and would not generate a deceptive log message. There's some relevant discussion on this here: https://github.com/hashicorp/consul/issues/8078

That's why I opted against inlining the config_file_registration token fallback into aclTokenForServiceSync. Although it does leave me a question: why does deleteService try using the service token and then fallback to agent token, instead of unilaterally using the agent token for deregistrations?

File: agent/token/store.go (hunk 3 above)
Comment (pglass) [reply agent/token/store.go]: Ah, I meant to re-refactor the other `Update*Token` methods to use the `updateToken` function I added here.

I added that in https://github.com/hashicorp/consul/pull/15828/commits/f814544fc7172a41d384ba9cc2bd40407d4dbfa5

File: agent/local/state.go (hunk 2 above)
Comment (jkirschner-hashicorp) [reply agent/local/state.go:1380]: My understanding is that the `agent` token typically only has `node:write` on itself. The agent token generally wouldn't have `service:write` needed to deregister a service.

Comment (pglass) [reply agent/local/state.go:1380]: > My understanding is that the agent token typically only has node:write on itself. The agent token generally wouldn't have service:write needed to deregister a service.

Right, but the Consul servers will accept a deregistration if token contains node:write for the node containing that service. See https://github.com/hashicorp/consul/pull/5217 and https://github.com/hashicorp/consul/blob/275a0b8e7fa73296b3c1b27742fe4a44f0e62b13/agent/consul/catalog_endpoint.go#L441-L448

Comment (jkirschner-hashicorp) [reply agent/local/state.go:1380]: Ah, interesting!

Comment (pglass) [reply agent/local/state.go:1380]: > Although it does leave me a question: why does deleteService try using the service token and then fallback to agent token, instead of unilaterally using the agent token for deregistrations?

I've tested this.

The agent only falls back to the agent token if the service token is _unset_ for that particular service (i.e. `token` field is empty or absent from the service definition in any config files). It does not fallback to other tokens on failure to deregister (i.e. it will not try the deregistration with the service token and, on failure, then try the agent token).

If the original service token has been deleted from the servers, because the agent has stored that service token in its local state, it continues to use that original service token to deregister that service during each state sync - which will repeatedly fail each time the state sync is retried. This feels like a bit of a gotcha.

Basically, I'm weighing two options:

1. The existing behavior is "good", so we should have deleteService include the config_file_registration token in its list of fallbacks. If set, the config_file_registration would be used instead of the agent token. And if the config_file_registration token was deleted, then the deregistration would fail forever.
2. Or, the existing behavior isn't great, and it would be better for it to only use the agent token for service deregistrations because that will "just work" because of the node:write "bypass" for service deregistrations.

Thoughts @jkirschner-hashicorp?

Comment (jkirschner-hashicorp) [reply agent/local/state.go:1380]: Is it accurate to say that the agent basically isn't functional if it lacks a token with `node:write` on itself?

If so... Is there any downside to approach 2? If I understand correctly, approach 2 would always work assuming the node is still functional (has a token the `node:write`).

Is there a separate case where a service is being deregistered directly from the server agents rather than from the node that owns the service, in cases where the node no longer exists but the service was never deregistered (but needs to be cleaned up)?

Happy to have a quick Zoom about this tomorrow.

**Comment by pglass** [comment]
I've updated this, so I've requested a re-review @kisunji 

* Rename `config_file_registration` to `config_file_service_registration`
* Checks in config files also use the `config_file_service_registration` token. This is because checks can be inlined in service definitions and service-level checks do not need additional permissions, so it makes sense for service-level checks to be registered with this token as well.
* Updated the external docs
* Added changelog
* Rebased onto main

File: website/content/docs/agent/config/config-files.mdx [hunk 4]
```diff
@@ -914,6 +914,21 @@ Valid time units are 'ns', 'us' (or 'µs'), 'ms', 's', 'm', 'h'."
...
+
+      If an inline token is defined in the service or check definition, then the inline token is
+      used to register that service or check instead. If the `config_file_service_registration` token is not
+      defined and if a service or check has no inline token, then the agent uses the
+      [`default`](#acl_tokens_default) token to register the service or check.
+
+      This token needs permissions for all services and checks this agent needs to register. For
+      example, if there are multiple service definitions in configuration files, then the
+      `config_file_service_registration` token needs multiple `service:write` permissions in order for
+      the agent to register those services.
```
[the remark this answers is not in the document]
Comment (kisunji) [reply website/content/docs/agent/config/config-files.mdx:930]: On that note, what happens if a `config_file_service_registration` token has permissions for a partial set of services? Does it fail to write all services or does it skip only the service with the missing perm?

Could make the behavior clear in the docs here.

**Review by kisunji — COMMENTED** [review]
LGTM; left mostly docs suggestions

File: website/content/docs/agent/config/config-files.mdx [hunk 5]
```diff
@@ -914,6 +914,21 @@ Valid time units are 'ns', 'us' (or 'µs'), 'ms', 's', 'm', 'h'."
     - `agent_master` ((#acl_tokens_agent_master)) **Renamed in Consul 1.11 to
       [`acl.tokens.agent_recovery`](#acl_tokens_agent_recovery).**
 
+    - `config_file_service_registration` ((#acl_tokens_config_file_service_registration)) - The ACL token this agent
+      uses to register services and checks from [service definitions](/docs/discovery/services) and
+      [check definitions](/docs/discovery/checks) foudn in configuration files or in configuration
+      strings passed to the agent using the `-hcl` flag.
```
[the remark this answers is not in the document]
Comment (jkirschner-hashicorp) [reply website/content/docs/agent/config/config-files.mdx:920]: Does registration also happen on reload?

If so, we could do something like:
```suggestion
      [check definitions](/docs/discovery/checks) loaded by the agent on startup and reload.
```

[the remark this answers is not in the document]
Comment (jkirschner-hashicorp) [reply website/content/docs/agent/config/config-files.mdx:920]: nit: If we keep the `-hcl` flag description, would we also need to include `-json`?

[the remark this answers is not in the document]
Comment (pglass) [reply website/content/docs/agent/config/config-files.mdx:920]: From what I can tell, `-json` doesn't exist. The `consul agent` command only has `-hcl` for config fragments.

[the remark this answers is not in the document]
Comment (blake) [reply website/content/docs/agent/config/config-files.mdx:920]: The `-hcl` flag enables operators to specify agent configuration values on the CLI. There is currently no equivalent `-json` flag for allowing agent configuration to be provided in JSON format. If we wanted to support that, it would be a new feature that requires additional development.

[the remark this answers is not in the document]
Comment (jkirschner-hashicorp) [reply website/content/docs/agent/config/config-files.mdx:920]: I don't want to suggest we support that. For some reason I thought I had seen an invocation of a Consul agent with that recently, but it seems like I misremembered (e.g., perhaps it was a config file being passed in with JSON format).

[the remark this answers is not in the document]
Comment (pglass) [reply website/content/docs/agent/config/config-files.mdx:920]: I think I prefer to be more elaborate / specific here to help reduce confusion?

If we say "configuration passed at startup", I feel like "startup" leaves room for interpretation. Is sending an HTTP request passing configuration to the agent? Does that include if I send a service registration request while the agent is "starting up"? 

I wanted to be clear about what services/checks the token is specifically used for (those services/checks in files or `-hcl` config fragments).

File: agent/local/state_internal_test.go [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+			isLocal:           true,
+			expToken:          "token123",
+		},
+		"empty token when resource is not locally-defined": {
+			registrationToken: "token123",
+			addResource:       addResourceFn,
+		},
+	}
+	for name, c := range cases {
+		c := c
```
[the remark this answers is not in the document]
Comment (pglass) [reply agent/local/state_internal_test.go:62]: Oh, it is. I have a code editor snippet that always adds this. I can remove it

File: agent/config/config.go [hunk 7]
```diff
@@ -754,11 +754,12 @@ type ACL struct {
...
-	Replication       *string `mapstructure:"replication"`
-	AgentRecovery     *string `mapstructure:"agent_recovery"`
-	Default           *string `mapstructure:"default"`
-	Agent             *string `mapstructure:"agent"`
+	InitialManagement      *string `mapstructure:"initial_management"`
+	Replication            *string `mapstructure:"replication"`
+	AgentRecovery          *string `mapstructure:"agent_recovery"`
+	Default                *string `mapstructure:"default"`
+	Agent                  *string `mapstructure:"agent"`
+	ConfigFileRegistration *string `mapstructure:"config_file_service_registration"`
```
[the remark this answers is not in the document]
Comment (pglass) [reply agent/config/config.go:762]: Yes. With Golang's general style preference for shorter variables, I thought `ConfigFileRegistration` was already long and is clear enough.

File: agent/local/state.go [hunk 8]
```diff
@@ -275,25 +288,26 @@ func (l *State) addServiceLocked(service *structs.NodeService, token string) err
...
+		Service:          service,
+		Token:            token,
+		IsLocallyDefined: isLocal,
 	})
 	return nil
 }
 
 // AddServiceWithChecks adds a service entry and its checks to the local state atomically
 // This entry is persistent and the agent will make a best effort to
 // ensure it is registered
```
[the remark this answers is not in the document]
Comment (pglass) [reply agent/local/state.go:300]: Yep. I updated this.

fwiw, I used "local" in order match [`ConfigSourceLocal`](https://github.com/hashicorp/consul/blob/c708e4d5dd303dd992b37dfa42f892b5b98510fc/agent/agent.go#L114).

File: website/content/docs/agent/config/config-files.mdx (hunk 4 above)
[the remark this answers is not in the document]
Comment (pglass) [reply website/content/docs/agent/config/config-files.mdx]: I rewrote this using two named services "A" and "B" as an example. I also elaborated a bit more on the failure case (maybe too much?). Let me know what you think!

**Review by skpratt — APPROVED** [review]
